File: latest/week3/workshop8_m10/W08_M10_demo.py
"""W08-M10: Lab — Build an Interactive Dashboard — Python Template — UNYT"""
import os; os.makedirs("output", exist_ok=True)

# Complete Dash lab template saved as standalone app
lab_app = '''
from dash import Dash, html, dcc, Input, Output, dash_table
import dash_bootstrap_components as dbc
import pandas as pd, plotly.express as px, plotly.graph_objects as go

# ── STEP 1: Global palette (M06) ────────────────────────
PALETTE = {"Movie": "#1565C0", "TV Show": "#E53935"}

# ── STEP 2: Load data ───────────────────────────────────
nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year
nf["primary_country"] = nf["country"].str.split(",").str[0].str.strip()
countries = sorted(nf["primary_country"].dropna().unique())

app = Dash(__name__, external_stylesheets=[dbc.themes.FLATLY])
server = app.server  # Required for Render deployment

# ── STEP 3: Helpers ──────────────────────────────────────
def make_kpi(title, value, color):
    fmt = f"{value:,}" if isinstance(value, int) else f"{value:,.0f}" if value > 100 else f"{value:.1f}%"
    return dbc.Card(dbc.CardBody([
        html.H6(title, className="text-muted mb-1", style={"fontSize": "0.75rem"}),
        html.H3(fmt, className="fw-bold mb-0", style={"color": color}),
    ]), className="text-center shadow-sm")

def filter_data(df, content_type, years, country):
    df = df.dropna(subset=["year_added"]).query(
        f"{years[0]} <= year_added <= {years[1]}")
    if content_type != "All":
        df = df[df["type"] == content_type]
    if country != "All":
        df = df[df["primary_country"] == country]
    return df

# ── STEP 4: Layout (Few's grid) ─────────────────────────
app.layout = dbc.Container([
    html.H3("Netflix Interactive Dashboard — W08 Lab",
        className="text-danger fw-bold mt-2 mb-2"),

    # Filter bar (Shneiderman: Filter)
    dbc.Row([
        dbc.Col(dcc.Dropdown(id="type-dd",
            options=[{"label":t,"value":t} for t in ["All","Movie","TV Show"]],
            value="All", clearable=False), width=2),
        dbc.Col(dcc.RangeSlider(id="year-sl", min=2015, max=2021, step=1,
            value=[2015,2021],
            marks={y:str(y) for y in range(2015,2022)}), width=5),
        dbc.Col(dcc.Dropdown(id="country-dd",
            options=[{"label":"All","value":"All"}]+
                [{"label":c,"value":c} for c in countries],
            value="All", clearable=False), width=3),
        dbc.Col(html.Button("Reset",id="reset",n_clicks=0,
            className="btn btn-outline-secondary btn-sm mt-1"),width=2),
    ], className="mb-2 p-2",
       style={"background":"#f9f9f9","borderRadius":"8px"}),

    # KPI row (Shneiderman: Overview)
    dbc.Row(id="kpi-row", className="mb-2"),

    # Primary + Secondary
    dbc.Row([
        dbc.Col(dcc.Graph(id="trend", style={"height":"300px"}), width=6),
        dbc.Col(dcc.Graph(id="countries", style={"height":"300px"}), width=6),
    ], className="mb-2"),

    # Detail row
    dbc.Row([
        dbc.Col(dcc.Graph(id="type-pie", style={"height":"260px"}), width=4),
        dbc.Col(dcc.Graph(id="ratings", style={"height":"260px"}), width=4),
        dbc.Col(dash_table.DataTable(id="table",
            columns=[{"name":c,"id":c} for c in
                ["title","type","year_added","rating","duration"]],
            page_size=8, sort_action="native", filter_action="native",
            style_table={"overflowX":"auto","maxHeight":"260px"},
            style_cell={"fontSize":"11px","padding":"4px"}), width=4),
    ]),
], fluid=True)

# ── STEP 5: Main callback (single source of truth) ──────
@app.callback(
    Output("trend","figure"), Output("countries","figure"),
    Output("type-pie","figure"), Output("ratings","figure"),
    Output("table","data"), Output("kpi-row","children"),
    Input("type-dd","value"), Input("year-sl","value"),
    Input("country-dd","value"))
def update_all(content_type, years, country):
    df = filter_data(nf, content_type, years, country)

    # Trend
    yr = df.groupby(["year_added","type"]).size().reset_index(name="n")
    fig_trend = px.line(yr, x="year_added",y="n",color="type",
        color_discrete_map=PALETTE, markers=True, title="Yearly Additions by Type")
    fig_trend.update_layout(template="plotly_white",hovermode="x unified",
        legend=dict(orientation="h",y=-0.15),margin=dict(t=40,b=40))

    # Countries
    top10 = df.dropna(subset=["primary_country"])["primary_country"]\\
        .value_counts().head(10).sort_values()
    fig_countries = px.bar(x=top10.values,y=top10.index,orientation="h",
        title="Top 10 Countries",labels={"x":"Titles","y":""})
    fig_countries.update_traces(marker_color="#1565C0",marker_opacity=0.7)
    fig_countries.update_layout(template="plotly_white",margin=dict(t=40,b=20),
        yaxis={"autorange":"reversed"})

    # Type pie
    type_ct = df["type"].value_counts().reset_index()
    type_ct.columns = ["type","count"]
    fig_pie = px.pie(type_ct,names="type",values="count",hole=0.4,
        color="type",color_discrete_map=PALETTE,title="Type Split")
    fig_pie.update_layout(showlegend=False,margin=dict(t=40,b=20))

    # Ratings
    ratings = df.dropna(subset=["rating"])["rating"].value_counts().head(8).sort_values()
    fig_ratings = px.bar(x=ratings.values,y=ratings.index,orientation="h",
        title="Top Ratings",labels={"x":"Count","y":""})
    fig_ratings.update_traces(marker_color="#2E7D32",marker_opacity=0.7)
    fig_ratings.update_layout(template="plotly_white",margin=dict(t=40,b=20))

    # Table
    table_data = df[["title","type","year_added","rating","duration"]]\\
        .head(200).to_dict("records")

    # KPIs
    total=len(df); movies=(df.type=="Movie").sum()
    tvshows=(df.type=="TV Show").sum(); nc=df["primary_country"].nunique()
    kpis = dbc.Row([
        dbc.Col(make_kpi("Total Titles",total,"#333"),width=3),
        dbc.Col(make_kpi("Movies",movies,"#1565C0"),width=3),
        dbc.Col(make_kpi("TV Shows",tvshows,"#E53935"),width=3),
        dbc.Col(make_kpi("Countries",nc,"#7B1FA2"),width=3)])

    return fig_trend, fig_countries, fig_pie, fig_ratings, table_data, kpis

# Cross-filter: click country → update dropdown (Shneiderman: Relate)
@app.callback(Output("country-dd","value"),
    Input("countries","clickData"),prevent_initial_call=True)
def on_click(click_data):
    if click_data and click_data.get("points"):
        return click_data["points"][0].get("y","All")
    return "All"

# Reset (Shneiderman: History)
@app.callback(Output("type-dd","value"),Output("year-sl","value"),
    Input("reset","n_clicks"),prevent_initial_call=True)
def reset(n): return "All",[2015,2021]

if __name__=="__main__":
    app.run(debug=True,port=8050)
'''
with open("output/lab_dashboard.py","w") as f: f.write(lab_app)
print("Lab template saved: output/lab_dashboard.py")
print("Run: python output/lab_dashboard.py → http://localhost:8050")

# ── SHNEIDERMAN + FEW AUDITS ────────────────────────────
print("\n=== SHNEIDERMAN AUDIT ===")
audit = [
    ("1. Overview first", "KPI row (4 cards)", "✓"),
    ("2. Zoom", "plotly scroll-zoom on all charts", "✓"),
    ("3. Filter", "Type dropdown + Year slider + Country dropdown", "✓"),
    ("4. Details on demand", "Hover tooltips on all plotly charts", "✓"),
    ("5. Relate", "Click country bar → filter all panels", "✓"),
    ("6. History", "Reset button restores defaults", "✓"),
    ("7. Extract", "DataTable with sort/search/filter", "✓"),
]
for step, component, status in audit:
    print(f"  {status} {step}: {component}")

print("\n=== FEW'S RULES AUDIT ===")
few = [
    ("1. One screen", "No scrolling at 1920×1080"),
    ("2. KPIs at top", "4 cards in first row"),
    ("3. ≤7 charts", "5 panels + 1 table = 6"),
    ("4. Consistent colour", "PALETTE applied globally"),
    ("5. Filter bar", "Top row with 3 inputs + reset"),
    ("6. Descriptive titles", "All titles state the metric"),
]
for rule, status in few:
    print(f"  ✓ {rule}: {status}")

print("\nAll W08-M10 Python outputs saved")

File: latest/week3/workshop8_m10/W08_M10_demo.R
# ============================================================
# Workshop 8 — Module 10: Lab — Build an Interactive Dashboard
# TEMPLATE R Script (Shiny) — UNYT Tirana
# ============================================================
# This is a TEMPLATE students adapt for their chosen dataset.
# It demonstrates the complete pipeline on Netflix.
library(tidyverse); library(shiny); library(plotly); library(DT)

# ═══════════════════════════════════════════════════════
# STEP 1: Define global palette (M06)
# ═══════════════════════════════════════════════════════
PALETTE <- c(Movie = "#1565C0", "TV Show" = "#E53935")

# STEP 2: Load data
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added)),
    year_added = year(date_added),
    primary_country = str_trim(str_extract(country, "^[^,]+")),
    duration_num = parse_number(duration))
countries <- sort(unique(na.omit(nf$primary_country)))

# ═══════════════════════════════════════════════════════
# STEP 3: UI — Few's grid layout (M06)
# ═══════════════════════════════════════════════════════
ui <- fluidPage(
  tags$head(tags$style(HTML("
    .kpi{text-align:center;padding:10px;border:1px solid #eee;border-radius:8px;background:white;}
    .kpi h3{margin:0;} .kpi p{margin:2px 0;font-size:11px;color:#888;}
  "))),

  # Title bar
  titlePanel(div(style="color:#C62828;","Netflix Interactive Dashboard — W08 Lab")),

  # Filter bar (M01: Filter step)
  fluidRow(style="background:#f9f9f9;padding:10px;margin-bottom:8px;border-radius:8px;",
    column(3, selectInput("type","Type:", c("All","Movie","TV Show"))),
    column(4, sliderInput("years","Years:", 2015, 2021, c(2015, 2021), sep="")),
    column(3, selectInput("country","Country:", c("All", countries))),
    column(2, br(), actionButton("reset","Reset", class="btn-outline-secondary btn-sm"))),

  # KPI row (M01: Overview step; M06: Rule 2)
  fluidRow(uiOutput("kpis")),

  # Primary + Secondary (M06: Rule 3)
  fluidRow(
    column(6, plotlyOutput("trend", height="300px")),
    column(6, plotlyOutput("country_bar", height="300px"))),

  # Detail row
  fluidRow(
    column(4, plotlyOutput("type_pie", height="260px")),
    column(4, plotlyOutput("rating_bar", height="260px")),
    column(4, DTOutput("table")))
)

# ═══════════════════════════════════════════════════════
# STEP 4: Server — single reactive + renders (M04)
# ═══════════════════════════════════════════════════════
server <- function(input, output, session) {

  # Single reactive: feeds ALL panels
  filtered <- reactive({
    df <- nf |> filter(!is.na(year_added),
      year_added >= input$years[1], year_added <= input$years[2])
    if (input$type != "All") df <- filter(df, type == input$type)
    if (input$country != "All") df <- filter(df, primary_country == input$country)
    df
  })

  # KPIs (M06: value + delta)
  output$kpis <- renderUI({
    n <- nrow(filtered()); m <- sum(filtered()$type == "Movie", na.rm=TRUE)
    tv <- sum(filtered()$type == "TV Show", na.rm=TRUE)
    nc <- n_distinct(filtered()$primary_country, na.rm=TRUE)
    make <- function(val, label, col) {
      column(3, div(class="kpi",
        h3(style=paste0("color:",col), format(val, big.mark=",")), p(label)))
    }
    fluidRow(make(n,"Total Titles","#333"), make(m,"Movies","#1565C0"),
      make(tv,"TV Shows","#E53935"), make(nc,"Countries","#7B1FA2"))
  })

  # Trend (M02: ggplotly + unified hover)
  output$trend <- renderPlotly({
    yr <- filtered() |> count(year_added, type)
    p <- ggplot(yr, aes(x=year_added, y=n, color=type,
        text=paste0(type,": ",n," (",year_added,")"))) +
      geom_line(linewidth=1.2) + geom_point(size=2) +
      scale_color_manual(values=PALETTE) +
      theme_minimal(base_size=9) + theme(legend.position="bottom") +
      labs(title="Yearly Additions by Type", x=NULL, y="Titles", color=NULL)
    ggplotly(p, tooltip="text") |> layout(hovermode="x unified")
  })

  # Country bar (M02: source for cross-filter)
  output$country_bar <- renderPlotly({
    top10 <- filtered() |> filter(!is.na(primary_country)) |>
      count(primary_country, sort=TRUE) |> slice_max(n, n=10) |>
      mutate(primary_country=fct_reorder(primary_country, n))
    p <- ggplot(top10, aes(x=primary_country, y=n,
        text=paste0(primary_country,": ",n))) +
      geom_col(fill="#1565C0", alpha=0.7, width=0.6) +
      coord_flip() + theme_minimal(base_size=9) +
      labs(title="Top 10 Countries", x=NULL, y="Titles")
    ggplotly(p, tooltip="text", source="country_src") |>
      layout(clickmode="event")
  })

  # Cross-filter: click country → update dropdown (M01: Relate)
  observeEvent(event_data("plotly_click", source="country_src"), {
    click <- event_data("plotly_click", source="country_src")
    if (!is.null(click)) updateSelectInput(session, "country", selected=click$y)
  })

  # Type pie/donut (variety of chart types)
  output$type_pie <- renderPlotly({
    type_ct <- filtered() |> count(type)
    plot_ly(type_ct, labels=~type, values=~n, type="pie",
      marker=list(colors=c("#1565C0","#E53935")),
      textinfo="label+percent", hole=0.4) |>
      layout(title=list(text="Type Split", font=list(size=12)),
        showlegend=FALSE, margin=list(t=40))
  })

  # Rating bar
  output$rating_bar <- renderPlotly({
    ratings <- filtered() |> filter(!is.na(rating)) |>
      count(rating, sort=TRUE) |> slice_max(n, n=8) |>
      mutate(rating=fct_reorder(rating, n))
    p <- ggplot(ratings, aes(x=rating, y=n, text=paste0(rating,": ",n))) +
      geom_col(fill="#2E7D32", alpha=0.7, width=0.6) +
      coord_flip() + theme_minimal(base_size=9) +
      labs(title="Top Ratings", x=NULL, y="Count")
    ggplotly(p, tooltip="text")
  })

  # Data table (M01: Extract step)
  output$table <- renderDT({
    filtered() |> select(title, type, year_added, rating, duration) |>
      head(200) |>
      datatable(options=list(pageLength=8, scrollX=TRUE, dom="ftip"),
        filter="top", rownames=FALSE)
  })

  # Reset (M01: History step)
  observeEvent(input$reset, {
    updateSelectInput(session, "type", selected="All")
    updateSliderInput(session, "years", value=c(2015, 2021))
    updateSelectInput(session, "country", selected="All")
  })
}

# shinyApp(ui, server)

# ═══════════════════════════════════════════════════════
# SHNEIDERMAN AUDIT
# ═══════════════════════════════════════════════════════
cat("\n══ SHNEIDERMAN AUDIT ══\n")
audit <- tibble(
  step = c("1. Overview first","2. Zoom","3. Filter","4. Details on demand",
    "5. Relate","6. History","7. Extract"),
  component = c("KPI row (4 cards)","plotly scroll-zoom on all charts",
    "Type dropdown + Year slider + Country dropdown",
    "Hover tooltips on all plotly charts",
    "Click country bar → filter all panels",
    "Reset button restores defaults",
    "DT table with sort/search/filter"),
  implemented = rep("✓", 7))
print(audit, n=7)

cat("\n══ FEW'S RULES AUDIT ══\n")
few <- tibble(
  rule = c("1. One screen","2. KPIs at top","3. ≤7 charts","4. Consistent colour",
    "5. Filter sidebar","6. Descriptive titles"),
  status = c("✓ No scrolling at 1920x1080","✓ 4 KPI cards in first row",
    "✓ 5 panels + 1 table = 6","✓ PALETTE applied globally",
    "✓ Filter bar at top","✓ All titles descriptive"))
print(few, n=6)

cat("\n── Lab Template Complete ──\n")
cat("Run: shinyApp(ui, server)\n")
